actionview/lib/action_view/helpers/form_helper.rb

```diff
@@ -1914,14 +1914,19 @@ def fields_for(record_name, record_object = nil, fields_options = {}, &block)
         fields_options[:namespace] = options[:namespace]
         fields_options[:parent_builder] = self
 
-        case record_name
-        when String, Symbol
-          if nested_attributes_association?(record_name)
-            return fields_for_with_nested_attributes(record_name, record_object, fields_options, block)
-          end
-        else
-          record_object = record_name.is_a?(Array) ? record_name.last : record_name
-          record_name   = model_name_from_record_or_class(record_object).param_key
+        if !record_name.is_a?(String) && !record_name.is_a?(Symbol)
+          record_object = record_name
```

Comment:
Basically you may move usage of `record_name` to the ` object_for_model_name = record_object.last`, instead of `record_object`.